src/components/data-list.jsx

```diff
@@ -1,57 +1,61 @@
-import React, { useState, useRef, useEffect } from 'react';
+import React, { Fragment } from 'react';
 import PropTypes from 'prop-types';
-import { v1 } from 'uuid';
-import {
-  checkLoadMoreItems,
-  checkOnDataLoad,
-  checkOnLoadMoreItems,
-} from './utils/scroll-data-loader';
-import '../styles/components/data-list.scss';
+import withDataLoader from './data-loader';
+import Card from './card';
 
-const loadingMessageId = v1();
-
-const DataList = ({ onLoadMoreItems, hasMoreData, items }) => {
-  const [loading, setLoading] = useState(false);
-  const [loadMoreItems, setLoadMoreItems] = useState(false);
-  const ref = useRef();
-
-  useEffect(
-    () => checkOnDataLoad(hasMoreData, setLoading, onLoadMoreItems, setLoadMoreItems, ref),
-    [items.length],
-  );
-
-  useEffect(() => checkOnLoadMoreItems(loadMoreItems, setLoading, onLoadMoreItems), [
-    loadMoreItems,
-  ]);
-
-  return (
-    <div className="data-list__wrapper">
-      <div
-        className="data-list__scroll-container"
-        onScroll={() => checkLoadMoreItems(loading, hasMoreData, ref, setLoadMoreItems)}
-        ref={ref}
+const DataList = ({
+  data, idKey, selectable, selected, onSelect, dataRenderer,
+}) => (
+  <Fragment>
+    {data.map(({ [idKey]: id, ...content }) => (
+      <Card
+        key={id}
+        {...{
+          selectable,
+          selected: !!selected[id],
+          onSelect: () => onSelect(id),
+        }}
       >
-        {items}
-        {loading && <h4 key={loadingMessageId}>Loading...</h4>}
-      </div>
-    </div>
-  );
-};
+        {dataRenderer(content)}
+      </Card>
+    ))}
+  </Fragment>
+);
 
 DataList.propTypes = {
   /**
-   * Callback to request more items if user scrolled to the bottom of the scroll-container or if
-   * the scroll-container isn't scrollable yet because not enough items have been loaded yet.
+   * The data to be displayed
+   */
+  data: PropTypes.arrayOf(PropTypes.object).isRequired,
+  /**
+   * The name of an attribute in each of the data objects which serves as a unique ID
    */
-  onLoadMoreItems: PropTypes.func.isRequired,
+  idKey: PropTypes.string,
   /**
-   * A boolean to indicate that the parent has more items to provide.
+   * A callback function that is called whenever a user selects a row. The row ID is returned upon
+   * callback.
    */
-  hasMoreData: PropTypes.bool.isRequired,
+  onSelect: PropTypes.func,
   /**
-   * An array of JSX elements which will populate the scrollContainer.
+   * An object which indicates which rows have been selected by the user.
    */
-  items: PropTypes.arrayOf(PropTypes.element).isRequired,
+  selected: PropTypes.instanceOf(Object),
```

Comment:
Can the shape of `selected` be defined here or the problem is the dynamic keys?

Reply:
Dynamic keys but we can use `shapeOf` to check the values